marlowe-playground-client/src/Marlowe/Linter.purs

```diff
@@ -0,0 +1,154 @@
+module Marlowe.Linter
+  ( lint
+  , State(..)
+  , Position
+  , _holes
+  , _negativeDeposits
+  , _negativePayments
+  ) where
+
+import Prelude
+import Data.Array (catMaybes, fold, foldMap, (:))
+import Data.BigInteger (BigInteger)
+import Data.Lens (Lens', over, set, view)
+import Data.Lens.Iso.Newtype (_Newtype)
+import Data.Lens.Record (prop)
+import Data.Maybe (Maybe(..))
+import Data.Newtype (class Newtype)
+import Data.Symbol (SProxy(..))
+import Data.Tuple.Nested (type (/\), (/\))
+import Marlowe.Holes (Action(..), Case(..), Contract(..), Holes, Term(..), Value(..), getHoles, insertHole)
+import Text.Parsing.StringParser (Pos)
+
+type Position
+  = { start :: Pos, end :: Pos }
+
+newtype State
+  = State
+  { holes :: Holes
+  , negativePayments :: Array Position
+  , negativeDeposits :: Array Position
+  }
+
+derive instance newtypeState :: Newtype State _
+
+instance semigroupState :: Semigroup State where
+  append (State s1) (State s2) =
+    State
+      { holes: s1.holes <> s2.holes
+      , negativePayments: s1.negativePayments <> s2.negativePayments
+      , negativeDeposits: s1.negativeDeposits <> s2.negativeDeposits
+      }
+
+instance monoidState :: Monoid State where
+  mempty =
+    State
+      { holes: mempty
+      , negativePayments: mempty
+      , negativeDeposits: mempty
+      }
+
+_holes :: Lens' State Holes
+_holes = _Newtype <<< prop (SProxy :: SProxy "holes")
+
+_negativePayments :: Lens' State (Array Position)
+_negativePayments = _Newtype <<< prop (SProxy :: SProxy "negativePayments")
+
+_negativeDeposits :: Lens' State (Array Position)
+_negativeDeposits = _Newtype <<< prop (SProxy :: SProxy "negativeDeposits")
+
+lint :: Term Contract -> State
+lint = traverseContract mempty
```

Comment:
This makes me think that `traverseContract` should be called `lintContract`. I mean, this isn't a generalised traversal of a contract, right? It only has one possible use.

Reply:
I could use the `lint = go mempty where go = ...` pattern?